identity-account-storage/src/storage/stronghold.rs

```diff
@@ -273,6 +280,70 @@ impl Storage for Stronghold {
     self.vault(did).exists(location.into()).await.map_err(Into::into)
   }
 
+  async fn encrypt_data(
+    &self,
+    did: &IotaDID,
+    data: Vec<u8>,
+    associated_data: Vec<u8>,
+    algorithm: &EncryptionAlgorithm,
+    private_key: &KeyLocation,
+    public_key: Option<PublicKey>,
+  ) -> Result<EncryptedData> {
+    let vault: Vault<'_> = self.vault(did);
+    match private_key.key_type {
+      KeyType::Ed25519 => aead_encrypt(&vault, algorithm, private_key, data, associated_data).await,
+      KeyType::X25519 => {
+        let shared_secret: SharedSecretLocation = SharedSecretLocation::random();
+        diffie_hellman(
+          &vault,
+          private_key,
+          public_key
+            .ok_or_else(|| Error::InvalidPublicKey("missing second party public key".to_owned()))?
+            .as_ref()
+            .try_into()
+            .map_err(|_| Error::InvalidPublicKey(format!("expected type: [u8, {}]", X25519::PUBLIC_KEY_LENGTH)))?,
+          &shared_secret,
+        )
+        .await?;
+        let encrypted_data: Result<EncryptedData> =
+          aead_encrypt(&vault, algorithm, &shared_secret, data, associated_data).await;
```

Comment:
I'm unsure how common or safe using the shared secret directly as the encryption key is. Most resources specify using a key-derivation function on the shared secret prior to using it for encryption. However, https://www.rfc-editor.org/rfc/rfc7518#section-4.5 does refer to using the shared secret directly, with security considerations regarding key lifetimes.

Basically, developers should be able to specify the key-wrapping algorithm used too.

Reply:
[Section 2.1 of the AEAD Interface spec](https://datatracker.ietf.org/doc/html/rfc5116#section-2.1) states that the encryption key:
> MUST be generated in a way that is uniformly random or pseudorandom.

Which is NOT the case for the shared secret produced by ECDH (as stated in for instance  [Section 3.3 of RFC 5869](https://datatracker.ietf.org/doc/html/rfc5869#section-3.3) and elsewhere). Hence we should probably indeed use a KDF (key-derivation function) on the shared secret. 

We could consider adding an additional optional parameter to this method specifying the KDF to use (possibly by introducing an enum of acceptable key-derivation functions (maybe with the possibility for the user to add a salt?)). Alternatively if there is some sufficiently common and recommended standard KDF to be used with ECDH we could decide to only support that.

Reply:
Thanks for the confirmation and sources.

To elaborate from a Slack conversation, [RFC 7518 Section 4.6](https://datatracker.ietf.org/doc/html/rfc7518#section-4.6), which defines `ECDH-ES`, specifies that the "Concat KDF" (which I believe uses a SHA digest per NIST recommendations) must be used to produce the content encryption key in both "Direct Key Agreement Mode" and "Key Wrapping Mode" (even though the wording of "key agreement result" is a bit misleading in my opinion):

> This section defines the specifics of key agreement with Elliptic
 Curve Diffie-Hellman Ephemeral Static [[RFC6090](https://datatracker.ietf.org/doc/html/rfc6090)], in combination with
  the **Concat KDF**, as defined in Section 5.8.1 of [[NIST.800-56A](https://datatracker.ietf.org/doc/html/rfc7518#ref-NIST.800-56A)].
>  The key agreement result can be used in one of two ways:
> 
> 1.  directly as the Content Encryption Key (CEK) for the "enc"
>    algorithm, in the Direct Key Agreement mode, or
> 
> 2.  as a symmetric key used to wrap the CEK with the "A128KW",
>    "A192KW", or "A256KW" algorithms, in the Key Agreement with Key
>    Wrapping mode.
> 
> A new ephemeral public key value MUST be generated for each key
> agreement operation.
> 
> **In Direct Key Agreement mode, the output of the Concat KDF MUST be a**
> **key of the same length as that used by the "enc" algorithm.** [...]

So basically without "Concat KDF" we would not technically have a spec-compliant ECDH-ES algorithm, which is a problem for interoperability if other libraries cannot decrypt it. (Aside: we definitely need to use their test vectors here.)

So we're stuck until https://github.com/iotaledger/stronghold.rs/issues/339 is resolved.

Reply:
Opened iotaledger/stronghold#344 to request the concat KDF.